bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/InterleavedLedgerStorage.java

```diff
@@ -465,6 +437,9 @@ public void onRotateEntryLog() {
         // TODO: we could consider remove checkpointSource and checkpointSouce#newCheckpoint
         // later if we provide kind of LSN (Log/Journal Squeuence Number)
         // mechanism when adding entry. {@link https://github.com/apache/bookkeeper/issues/279}
-        checkpointHolder.setNextCheckpoint(checkpointSource.newCheckpoint());
+        if (null != checkpointer) {
```

Comment:
in which case it can be null? and is it ok to be null?

Reply:
SortedLedgerStorage is setting it to null. The trigger of checkpoint for sortedledger storage should happen after flushing memtables, not when entrylog is rolled.

Reply:
1) it is very confusing that you added 'checkpointer'  variable both in InterleavedLedgerStorage and its subclass - SortedLedgerStorage. It is easy to get lost or misunderstand.

2.a) I'm not convinced if it is right approach to have different behavior depending on the type of LedgerStorage. In the case of InterleavedLedgerStorage, in the event of 'onRotateEntryLog' you trigger checkpoint but in the case of SortedLedgerStorage it is noop (you swallow it). 
2.b) If you dont agree with what I mentioned in 2.a), then I would like to ask is it necessary to have EntryLogListener.onRotateEntryLog? cannt we have something similar to CacheCallback.onSizeLimitReached in InterleavedLedgerStorage also and get rid of EntryLogListener?

Reply:
it is very confusing that you added 'checkpointer' variable both in InterleavedLedgerStorage and its subclass - SortedLedgerStorage. It is easy to get lost or misunderstand.

2.a) I'm not convinced if it is right approach to have different behavior depending on the type of LedgerStorage. In the case of InterleavedLedgerStorage, in the event of 'onRotateEntryLog' you trigger checkpoint but in the case of SortedLedgerStorage it is noop (you swallow it).

2.b) If you dont agree with what I mentioned in 2.a), then I would like to ask is it necessary to have EntryLogListener.onRotateEntryLog? cannt we have something similar to CacheCallback.onSizeLimitReached in InterleavedLedgerStorage also and get rid of EntryLogListener?

Reply:
1) it is very confusing that you added 'checkpointer' variable both in InterleavedLedgerStorage and its subclass - SortedLedgerStorage. It is easy to get lost or misunderstand.

2.a) I'm not convinced if it is right approach to have different behavior depending on the type of LedgerStorage. In the case of InterleavedLedgerStorage, in the event of 'onRotateEntryLog' you trigger checkpoint but in the case of SortedLedgerStorage it is noop (you swallow it).

2.b) If you dont agree with what I mentioned in 2.a), then I would like to ask is it necessary to have EntryLogListener.onRotateEntryLog? cannt we have something similar to CacheCallback.onSizeLimitReached in InterleavedLedgerStorage also and get rid of EntryLogListener?

Reply:
@reddycharan 

the logic here is to have each ledger storage control when they want to checkpoint. when ledger storage wants to do checkpoint, it issues a checkpoint request to checkpointer to ask it starts checkpointing. the idea was originated when we introduced checkpoint mechanism to avoid performance penalty. This change was just reverting the behavior the original checkpoint mechanism I introduced 3 years ago, which is proven to be correct and has been running at Twitter production for 3 years. Hope this clarifies 2. a)

the question 1 is unfortunately to be avoided (without changing a lot of code), if we follows the principle letting each ledger storage control when they want to checkpoint, because SortedLedgerStorage inherits from InterleavedLedgerStorage.

regarding 2. b), the idea for checkpointing sorted ledger storage is we only checkpoint sorted ledger storage, when the interleave ledger storage moving on to a new entry log file (which is effectively entry log file rolling). We don't want to checkpoint happens on a file that memtable will potentially flushing entries to. That hurts performance. so in this sense, the checkpoint for sorted ledger storage only happens after a memtable is flushed to entry log file and the entry log file is rotated to a new file.

The implementation here is not elegant as Ivan has raised the concerns 3 years ago. However, it turns out this implementation provides much efficient and correct behavior, which has been verified in twitter production for 3 years. I would suggest us going with this change and doing a refactor later. It is much safer from my experience.

Reply:
"the logic here is to have each ledger storage control when they want to checkpoint." Yes, I understood that and I agree in principle what you are doing.
1) I'm merely questioning the choice of variable names. Unless there is a reason, why do we have to get into variable hiding/shadowing. Why cannt we just name them differently say ilsCheckpointer in InterleavedLedgerStorage and slsCheckpointer in SortedLedgerStorage. Anyhow those variables are already private.

Reply:
oh, regarding variable -- I don't think there is any specific reasons about renaming. the code was added 3 years ago. I was just following the original path to revert the checkpoint logic back. since they are private, I don't think they will conflict with each other, so I didn't rename them. I am fine with renaming them if you feel strong about that.

Reply:
Just to add, we shouldn't go into a big refactor until the yahoo changes are in, because the yahoo changes will pull in a whole other ledger storage.

Reply:
@sijie I think it would be much better and cleaner if you have protected 'checkpointer' variable in InterleavedLedgerStorage and remove 'checkpointer' variable from SortedLedgerStorage. Then override onRotateEntryLog method in SLS and make it noop. So this way it would be easier to get what each class is doing and when they are calling startCheckpoint. The current handling is sneaky and one would wonder why we are having this null check and why we are doing variable hiding/shadowing.